fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeCheckerV2.java

```diff
@@ -495,32 +490,112 @@ static void checkOperateLoadPrivilege(ConnectContext context, String dbName, Str
+        public Void visitQueryStatement(QueryStatement statement, ConnectContext context) {
+            Map<TableName, Table> allTouchedTables = AnalyzerUtils.collectAllTableAndView(statement);
+            for (Map.Entry<TableName, Table> tableToBeChecked : allTouchedTables.entrySet()) {
+                TableName tableName = tableToBeChecked.getKey();
+                Table table = tableToBeChecked.getValue();
+
+                if (table instanceof View) {
+                    if (!PrivilegeActions.checkViewAction(context, tableName.getDb(), tableName.getTbl(), PrivilegeType.SELECT)) {
+                        ErrorReport.reportSemanticException(ErrorCode.ERR_PRIVILEGE_ACCESS_TABLE_DENIED,
+                                context.getQualifiedUser(), tableName);
+                    }
+                } else if (table instanceof SchemaTable && ((SchemaTable) table).isBeSchemaTable()) {
+                    checkStmtOperatePrivilege(context);
+                } else if (table.isMaterializedView()) {
+                    checkMvAction(context, tableName, PrivilegeType.SELECT);
+                } else {
+                    checkTableAction(context, tableName, PrivilegeType.SELECT);
+                }
+            }
+
             return null;
         }
 
+        // ------------------------------------------- DML Statement -------------------------------------------------------
+
         @Override
         public Void visitInsertStatement(InsertStmt statement, ConnectContext context) {
             // For table just created by CTAS statement, we ignore the check of 'INSERT' privilege on it.
             if (!statement.isForCTAS()) {
                 checkTableAction(context, statement.getTableName(), PrivilegeType.INSERT);
             }
+
+            visit(statement.getQueryStatement(), context);
+            return null;
+        }
+
+        @Override
+        public Void visitDeleteStatement(DeleteStmt statement, ConnectContext context) {
+            checkTableAction(context, statement.getTableName(), PrivilegeType.DELETE);
+
+            Map<TableName, Table> allTouchedTables = AnalyzerUtils.collectAllTableAndView(statement);
+            for (Map.Entry<TableName, Table> tableToBeChecked : allTouchedTables.entrySet()) {
+                TableName tableName = tableToBeChecked.getKey();
+                Table table = tableToBeChecked.getValue();
+
+                if (table instanceof View) {
+                    if (!PrivilegeActions.checkViewAction(context, tableName.getDb(), tableName.getTbl(), PrivilegeType.SELECT)) {
+                        ErrorReport.reportSemanticException(ErrorCode.ERR_PRIVILEGE_ACCESS_TABLE_DENIED,
+                                context.getQualifiedUser(), tableName);
+                    }
+                } else if (table instanceof SchemaTable && ((SchemaTable) table).isBeSchemaTable()) {
+                    checkStmtOperatePrivilege(context);
```

Comment:
we don't need to do this check here, schema tables are virtual tables, they cannot be deleted

Reply:
Only the target table checks the delete permission, and the other tables check the Select permission

Reply:
got it